### ISSUE TITLE
Apache and Nginx plugins should use reload rather than restart

### ISSUE THREAD

**Issue by jsha** [letsencrypt/letsencrypt#954]
letsencrypt-apache/letsencrypt_apache/configurator.py
>     .. todo:: Try to use reload instead. (This caused timing problems before)                                                                                                    

Our challenge methods are designed to be used without taking down a web server, so we should make sure we actually don't take down the web server to restart.

Note that Nginx has this issue too, noted in https://github.com/letsencrypt/letsencrypt/pull/950 and worked around.

We need a generic wait_for_challenge method that can be used for both plugins, which will repeatedly attempt to connect to the appropriate port on localhost and ask for the appropriate servername. If the loop times out after, say, five seconds, it should error out and say the configuration to add the challenge failed.

**Comment by jsha** [comment]
Copying from related PR:

-----jsha------
I think it's a little harder than it sounds, but I'd be interested in being proved wrong. My main concerns were:
(a) Can we be sure connecting to localhost is always the right thing? (probably yes)
(b) what method do I use to connect?

For (b), urllib seems to be the obvious answer, but we'd have to disable cert checking because it would be a snakeoil cert. However, if the new config hasn't loaded, we will get a default cert, so we won't get a nice failure like we wanted. Is there an easy to use connect method that will give me back the DNSNames of the provided cert regardless of cert validity?

-----kuba-----
(a) maybe just connect to the domain in question?
(b) you can use acme.crypto_util.probe_sni for that :)

(a) and (b) together: probe_sni(domain, socket.gethostbyname(domain), config.dvsni_port)

**Comment by joohoi** [comment]
Apache reload *should* make the new changes effective immediately, as it's using the vanilla apache graceful restart on most (all?) distros that it's available at (all?). Graceful restart *should* serve the existing connections with the old configuration, and the new ones with the new one. 

@jdkasten can you remember the testing process that led to the todo-comment?

I'll try to tinker with this in the coming days.

**Comment by jdkasten** [comment]
I haven't tested the reload since Apache 2.2 back when I initially wrote up
a demo.

Reload used to fail for me about 1/6th of the time.  The configurations
would be edited, reload would be called, but the challenges failed because
the actual server didn't seem to load in the new configuration.

I would manually have to restart the server in order for it to read the new
configuration (challenge configuration)

I also used to test it by running a looped wget bash script while the Let's
Encrypt client was running.  (Just to see any performance implications)  My
guess at the time was that some long lasting connections caused the stall.

When I changed it to restart (a very long time ago)  the failures
disappeared.  I had vowed to return to the problem...

So in other words, if it appears to be working with the simple change to
reload, that is fantastic.  I had problems a few years ago and hadn't
messed with it since.  Apache may have updated their scirpt, or perhaps I
misattributed a different problem with the restart command.

**Comment by jsha** [comment]
Note: there is almost certainly going to be a race condition on reload for both Nginx and Apache: the servers will take a non-zero amount of time to process the signal, parse the new config, and start serving with it. I did run into this race condition pretty reliably when working on the Nginx code. That's why I recommend adding an explicit loop check to wait until the new config is actually serving before proceeding with processing.

**Comment by jdkasten** [comment]
@jsha  Yes, the restart was implemented and //todo added... just to get
around the race condition temporarily.  I think the only viable solution is
to poll the server checking for the changes, it just requires a bit more
code.

IIRC, the reload timings (configuration changes) had a very high variance.

That being said, all webservers can reuse the same code to test for the
challenges.  (It can be added to the plugins directory within letsencrypt)

**Comment by bmw** [comment]
I've also been thinking about this lately and wanted to follow up here. At first, I thought we only needed to poll when setting up for challenges. The thinking was that we only really care that new configuration is in effect when boulder is about to perform DV.

Unfortunately, this doesn't work. According to the [Apache documentation](https://httpd.apache.org/docs/2.4/stopping.html#graceful), it's possible for the reload command to return 0 and Apache fails to reload the config. Therefore, we need to poll on every reload to make sure Apache was able to load the new config.

The polling approach is a bit tricky. The Apache plugin calls `restart` a number of times internally after making different changes. This means we'd need to poll for different changes depending on when `restart` is called.

Even worse, the Apache plugin implements `restart` as part of the `IInstaller` interface which is called a number of times by the rest of the code base. This means that `restart` needs to be able to figure out what changes to poll for on its own, based on nothing than its internal state. Perhaps we can do this by examining the state of Augeas, but this seems quite difficult.

**Comment by bmw** [comment]
Additionally, we can't always poll ourselves. Specifically, proxies/firewalls will be a problem.

**Comment by joohoi** [comment]
Something on the better side: I can't seem to be able to reproduce the issue with loading a new configuration with reload. The new config gets loaded every time.

I guess the problem previously could have had something to do with wget leaving connection open as default (without --no-http-keep-alive option)

However, as @jsha commented, the server takes time reloading the new configuration in any case, so some kind of delay needs to be in place. I haven't tested it with large amount of vhost configurations, so can't say for sure how the amount of different configurations affects the reload times. 

For smaller configuration sets the changes seem instant, at least with debian init scripts. I tested with:
```
sudo /etc/init.d/apache2 reload;curl http://test.local
```
And it produced the desired outcome every single time.

**Comment by joohoi** [comment]
@bmw, what kind of cases are you thinking with the comment about apache failing to reload the configuration? Something that we can't predict with configtest? At least in my tests both debian init script and apache2ctl errored out with return code 1. However all I really tested was configuration syntax errors.

**Comment by bmw** [comment]
@joohoi, so I actually don't know of a case where the problem occurs, but it is possible according to the Apache documentation I linked above. To quote the documentation of graceful restart (which has the same effect of calling the init script with `reload`):

>When you issue a restart, a syntax check is first run, to ensure that there are no errors in the configuration files. If your configuration file has errors in it, you will get an error message about that syntax error, and the server will refuse to restart. This avoids the situation where the server halts and then cannot restart, leaving you with a non-functioning server.

>This still will not guarantee that the server will restart correctly.

With that said, it appears `restart` has the same behavior and we currently don't do anything to protect against this case.

Perhaps we only have to do polling when setting up challenges. While not failproof, we could configure an additional vhost in the call to `perform`. We can then attempt to poll this vhost a fixed number of times. If we receive the expected response, we continue.

If we never receive the correct response, I think the best behavior is to attempt to continue here as well. The reason is it is possible we're unable to connect to ourselves for whatever reason but Apache has successfully loaded the new config.

**Comment by joohoi** [comment]
For how I see it, as such, we wouldn't gain anything from polling the changes, as the race condition seems to be attributed to the connection keepalive settings from the HTTP client side in the graceful restart (reload).

**Comment by jsha** [comment]
It'd be nice if we could track that down for sure. There's no hard reason why Apache and Nginx would not be able to start accepting new connections with a new config while old ones are still inflight. And in fact such a problem would effectively kill graceful reload for many sites that have very long keepalives.

**Comment by joohoi** [comment]
The keepalive problem usually effects only a subset of the child processes, and polling separate child processes is close to impossible in a coherent way. I suggest we leave the restart in place and introduce a --reload-only cli flag with appropriate warning to let the user choose whether to take a walk on the thin ice or not.

**Comment by joohoi** [comment]
To recap what's going on here (as a tl;dr and to clear my thoughts): 

  * We want to use reload instead of restart in both Apache and Nginx, but:
     * Reloading configuration (and it coming to an effect) takes undefined time, depending on the amount of configuration present etc.
     * There's an underlying keepalive issue with Apache child processes.
     * Polling via HTTP may not be feasible in some corner configuration cases.
     * Apache documentation states that there may be (is?) some corner cases where the new configuration will not be reloaded even though the return codes state so.
  * We can use apache2ctl -S to determine some parts of active vhost configuration, but it doesn't take the old child processes running and old configuration in to account.
  * Suggestion about letting the user decide if they want to use reload instead of restart, but let them know about the possible shortcomings.

**Comment by bmw** [comment]
> Reloading configuration (and it coming to an effect) takes undefined time, depending on the amount of configuration present etc.

I believe this is also true of `restart`. My understanding is that both `restart` and `reload` run a config test, send the Apache process a signal, and immediately return. Sure the signal and processing time is different, but both the `restart` and `reload` command return before the process is complete. The actual reload/restart takes an undefined amount of time.

> There's an underlying keepalive issue with Apache child processes.

So I can confirm from testing that Apache 2.4 can handle requests with the new and old configuration at the same time. In my tests, all new connections were handled with the new configuration, but I don't know that this always happens.

Due to the complexity of doing this right, I'm leaning towards either just adding a short call to `time.sleep` after reloading the config for challenges or kicking this out of the 1.0 milestone.

**Comment by jsha** [comment]
I think it's really important to have this for 1.0. I'm fine with a call to `sleep` as a short-term workaround. I'd suggest at least two seconds as a ballpark figure.
